core/store/src/trie/update.rs

```diff
@@ -241,29 +217,24 @@ impl<'a> TrieUpdateIterator<'a> {
         );
         let prospective_iter = state_update
             .prospective
-            .range(start_offset..)
+            .range(prefix.clone()..)
             .map(|(raw_key, key_value)| (raw_key.as_slice(), key_value.value.as_deref()));
         let overlay_iter = MergeIter {
             left: (Box::new(committed_iter) as Box<dyn Iterator<Item = _>>).peekable(),
             right: (Box::new(prospective_iter) as Box<dyn Iterator<Item = _>>).peekable(),
         }
         .peekable();
-        Ok(TrieUpdateIterator {
-            prefix: prefix.to_vec(),
-            end_offset,
-            trie_iter: trie_iter.peekable(),
-            overlay_iter,
-        })
+        Ok(TrieUpdateIterator { prefix, trie_iter: trie_iter.peekable(), overlay_iter })
     }
 }
 
 impl<'a> Iterator for TrieUpdateIterator<'a> {
     type Item = Result<Vec<u8>, StorageError>;
 
     fn next(&mut self) -> Option<Self::Item> {
-        let stop_cond = |key: &[u8], prefix: &[u8], end_offset: &Option<Vec<u8>>| {
+        let stop_cond = |key: &[u8], prefix: &[u8]| {
             // TODO(mina86): Figure out if starts_with check is still necessary.
```

Comment:
Unrelated to the PR, but it seems like the logic here can be simplified by removing `Ordering` and instead first unconditionally fetching the smaller key, and then doing the prefix check at the end as a guard.

Reply:
There’s in fact even better way to deal with this by creating a range iterator with upper bound set to smallest key larger then prefix which does not start with a prefix.  We’re already doing something similar in RocksDB.  See next_prefix function.  I have a couple WIP commits which improve this code some more.